File: warmup.py
from random import shuffle
import operator
from itertools import izip_longest,repeat


#a fuction uses divmod to find the amount of quarters,dimes,nickels and pennies in a specified amount

def change(amount):
    if amount < 0: raise ValueError, "amount cannot be negative"
    quarters,amount=divmod(amount,25)
    dimes,amount=divmod(amount,10)
    nickels,amount=divmod(amount,5)
    pennies=amount
    return (quarters,dimes,nickels,pennies)


#function that takes in a string and removes all quotes then returns a string surrounded by qoutes
def strip_quotes(string):
	string=string.replace("'","").replace('"','')
	return "'" + string +"'"

#function that takes a phrase and shuffles the characters around in a random order

def scramble(phrase):
	phrase_list=list(phrase)
	shuffle(phrase_list)
	new_phrase=''.join(phrase_list)
	return new_phrase


#a generator that takes in a number to be the range and finds exponents of 2 up to the range
def powers_of_two(number):
    for exponent in range(number):
       if 2**exponent>number:
          return
       else:
          yield 2**exponent


#a generator that takes in a string and produces an identical string letter by letter using the old string  
def prefixes(string):
    new_string=" "
    print new_string
    for elements in range(len(string)):
       new_string+=string[elements]
       yield new_string

#takes two lists and combines them through interleaving
def interleave(list1,list2):
  list3=reduce(operator.__add__, list(izip_longest(list1,list2,fillvalue="")))
  return [element for element in list3 if not element == '']
	
  
#takes in a list and duplicates each element of the list
def stutter(list_to_double):
   return reduce(operator.__add__,[list(repeat(element,2)) for element in list_to_double])

File: homework2/warmup.js
var shuffle = require('knuth-shuffle').knuthShuffle



//takes an amount an returns the amount of quarters dimes nickels and pennies
//in an array


exports.change=function (amount){
  if(amount<0){
    console.error("RangeError: amount cannot be negative")
    process.exit()
  }

   quarters=Math.floor(x/25)
   amount=amount % 25
   var dimes=Math.floor(x/10)
   amount= amount % 10
   var nickels=Math.floor(x/5)
   pennies=amount % 5
   return [quarters,dimes,nickels,pennies]

}

//removes quotes from a string

exports.stripQuotes=function (string){
	var quotes_removed= string.replace(/[\" \']/g,"")
    return quotes_removed
}


//scrambles a string so that each character has a random chance at being
//at any part of the string


exports.scramble=function(phrase){
   phrase=phrase.split("")
   phrase=shuffle(phrase)
   return phrase.reduce(function(a, b) {return a + b})
 }

//takes in a range and a function and gives exponents of 2 till the range is met

exports.powersOfTwo = function(range,print_function){
    var number =2
    var exponent=0
    var result= 0
    while (result<range){
      result=Math.pow(number,exponent)
      exponent++
      if (result>range){
        return
      }
      print_function(result)
    }
}


//takes in a string and a function then prints each character of the string
//one by one till the string is made

exports.prefixes=function(string, print_function){
     var adjusted_string=" "+string
     var result=adjusted_string.split("")
     var length=result.length
     var prefix_string=" "
     for(i=0;i<length;i++){
     print_function(prefix_string+=result[i])
     }

 }

//combines two arrays to make a new array that has the values of both the arrays
//in alternating order

exports.interleave= function(array1, array2){
  var array3= []
  if(array1.length>array2.length){
    length=array1.length
  }else if(array2.length>array1.length){
    length=array2.length
  }else{
    length=array1.length
  }

  for(i=0;i<length;i++){
    if(i<array1.length){
      array3.push(array1[i])
    }
    if(i<array2.length){
      array3.push(array2[i])
    }

  }

  
 return array3

}

//takes an array and interleaves it with itself


exports.stutter=function(array){
  return interleave(array,array)
